### PR TITLE
Adding Constant Check for FilterExec

### LINKED ISSUE #9612: Eliminate constant columns from sort requirements
### Is your feature request related to a problem or challenge?

I have a table sorted by `(date, ticker, time)` and would like to query data ordered by `(date, time)` assuming `ticker` is known to be constant. 

It seems that if `FilterExec` would emit singleton statistics if not for the following check failing:

https://github.com/apache/arrow-datafusion/blob/3c26e597aeacde0a5e6a51f30394d3d31c6acd96/datafusion/physical-plan/src/filter.rs#L127-L135

which ultimately boils down to this piece of code only allowing numeric types:

https://github.com/apache/arrow-datafusion/blob/3c26e597aeacde0a5e6a51f30394d3d31c6acd96/datafusion/physical-expr/src/intervals/utils.rs#L96-L111

I noticed that if we just add `DataType::Utf8` to the list, it seems to work for the example below (in the next section), so my uneducated mind wonders if we can just add more datatypes to it? 🤔

### Describe the solution you'd like

Given the following table:

```sql
CREATE EXTERNAL TABLE data (
    "date"   VARCHAR, 
    "ticker" VARCHAR, 
    "time"   VARCHAR,
) STORED AS PARQUET
WITH ORDER ("date", "ticker", "time")
LOCATION 'example/';

INSERT INTO data VALUES 
('2023', 'A', '0400'), 
('2023', 'A', '0401'), 
('2023', 'B', '0400');
```

I would like for the following query to be executed without a sort:

```sql
SELECT * FROM data 
WHERE ticker = 'A' 
ORDER BY "date", "time";
-- +---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
-- | plan_type     | plan                                                                                                                                                                                                                                                                                                                                                                                         -- |
-- +---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
-- | logical_plan  | Sort: data.date ASC NULLS LAST, data.time ASC NULLS LAST                                                                                                                                                                                                                                                                                                                                     -- |
-- |               |   Filter: data.ticker = Utf8("A")                                                                                                                                                                                                                                                                                                                                                            -- |
-- |               |     TableScan: data projection=[date, ticker, time], partial_filters=[data.ticker = Utf8("A")]                                                                                                                                                                                                                                                                                               -- |
-- | physical_plan | SortPreservingMergeExec: [date@0 ASC NULLS LAST,time@2 ASC NULLS LAST]                                                                                                                                                                                                                                                                                                                       -- |
-- |               |   SortExec: expr=[date@0 ASC NULLS LAST,time@2 ASC NULLS LAST]                                                                                                                                                                                                                                                                                                                               -- |
-- |               |     CoalesceBatchesExec: target_batch_size=8192                                                                                                                                                                                                                                                                                                                                              -- |
-- |               |       FilterExec: ticker@1 = A                                                                                                                                                                                                                                                                                                                                                               -- |
-- |               |         RepartitionExec: partitioning=RoundRobinBatch(16), input_partitions=1                                                                                                                                                                                                                                                                                                                -- |
-- |               |           ParquetExec: file_groups={1 group: [[Users/matthewcramerus/arrow-datafusion/datafusion-cli/example/bGDu3tzuQaHhgpGZ_0.parquet]]}, projection=[date, ticker, time], output_ordering=[date@0 ASC NULLS LAST, ticker@1 ASC NULLS LAST, time@2 ASC NULLS LAST], predicate=ticker@1 = A, pruning_predicate=ticker_min@0 <= A AND A <= ticker_max@1, required_guarantees=[ticker in (A)] |
-- |               |                                                                                                                                                                                                                                                                                                                                                                                              -- |
-- +---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```



### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #9612

### CODE REVIEW

**Comment by Lordworms** [comment]
A problem here is that when dealing with c in ('a', 'b', 'c') the parser would translate it into a BinaryOp::Eq and we could not directly do add the left part to constant.

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 1]
```diff
@@ -131,7 +131,7 @@ fn update_coalesce_ctx_children(
     coalesce_context.data = if children.is_empty() {
         // Plan has no children, it cannot be a `CoalescePartitionsExec`.
         false
-    } else if is_coalesce_partitions(&coalesce_context.plan) {
+    } else if is_coalesce_partitions(&coalesce_context.plan) || coalesce_context.data {
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs:134]: I think none of the changes in this file is necessary for this PR. I retracted changes in this file. All tests still pass

File: datafusion/physical-plan/src/filter.rs [hunk 2]
```diff
@@ -158,7 +160,55 @@ impl FilterExec {
...
+                    }
+                }
+                Ok(TreeNodeRecursion::Continue)
+            })
+            .expect("no way to return error during recursion");
+        if contains_or {
+            res_constants.clear();
+        }
+        res_constants
+    }
```
Comment (mustafasrepo) [inline datafusion/physical-plan/src/filter.rs:211]:
```suggestion
    fn extend_constants(
        input: &Arc<dyn ExecutionPlan>,
        predicate: &Arc<dyn PhysicalExpr>,
    ) -> Vec<Arc<dyn PhysicalExpr>> {
        let mut res_constants = Vec::new();
        let input_eqs = input.equivalence_properties();

        let conjunctions = split_conjunction(predicate);
        for conjunction in conjunctions {
            if let Some(binary) = conjunction.as_any().downcast_ref::<BinaryExpr>() {
                if binary.op() == &Operator::Eq {
                    if input_eqs.is_expr_constant(binary.left()) {
                        res_constants.push(binary.right().clone())
                    } else if input_eqs.is_expr_constant(binary.right()) {
                        res_constants.push(binary.left().clone())
                    }
                }
            }
        }
        res_constants
    }
```
I have refactored `extend_constants` helper to use existing util codes. I think, we can use this snippet instead.

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 3]
```diff
@@ -162,8 +162,18 @@ impl PhysicalOptimizerRule for EnforceSorting {
...
                 PlanWithCorrespondingCoalescePartitions::new_default(adjusted.plan);
+            plan_with_coalesce_partitions.children = plan_with_coalesce_partitions
+                .children
+                .into_iter()
+                .enumerate()
+                .map(|(id, mut child)| {
+                    child.data = data_vec[id];
+                    child
+                })
+                .collect();
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs:176]: I think these changes are still unnecessary for this PR. Other than this, this PR is ready to merge. Thanks @Lordworms for this Pr.

**Review by mustafasrepo — APPROVED** [review]
LGTM! Thanks @Lordworms again for this PR.

**Comment by Lordworms** [comment]
Don't know why the win test failed, is this due to  CI/CD issues?

**Comment by mustafasrepo** [comment]
> Don't know why the win test failed, is this due to CI/CD issues?

This unrelated to this Pr. There is an issue with windows runner. See issue: [9653](https://github.com/apache/arrow-datafusion/issues/9653)

**Comment by mustafasrepo** [comment]
> > Don't know why the win test failed, is this due to CI/CD issues?
> 
> This unrelated to this Pr. There is an issue with windows runner. See issue: [9653](https://github.com/apache/arrow-datafusion/issues/9653). This test is undeterministic (sometimes fails, sometimes passes). After CI completes I will re-rerun the test.

**Comment by Lordworms** [comment]
> > Don't know why the win test failed, is this due to CI/CD issues?
> 
> This unrelated to this Pr. There is an issue with windows runner. See issue: [9653](https://github.com/apache/arrow-datafusion/issues/9653)

Got it, thanks

**Pull request merged**